bpo-43950: Print columns in tracebacks (PEP 657)

ammaraskar opened the pull request:
This pull request implements the relevant bits in traceback printing to utilize the new `co_positions` and `PyCode_Addr2...`  APIs to print carets for column locations. 

**All the internal structures, and data representation are subjected to change in future PRs as we add optimizations.**

<!-- issue-number: [bpo-43950](https://bugs.python.org/issue43950) -->
https://bugs.python.org/issue43950
<!-- /issue-number -->

isidentical commented on Python/traceback.c:
```diff
@@ -501,8 +511,35 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent)
+    }
+    Py_ssize_t len = strlen(str);
+    if (offset > len) {
+        offset = len;
```
@ammaraskar let's use `Py_SAFE_DOWNCAST` here (getting warnings on windows)

isidentical commented on Python/traceback.c:
```diff
@@ -501,8 +511,35 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent)
+    }
+    Py_ssize_t size = PyUnicode_GET_LENGTH(text);
+    Py_DECREF(text);
+    return size;
```
here as well https://github.com/python/cpython/pull/26958/files#r663045408

pablogsal commented on Python/traceback.c:
```diff
@@ -501,8 +511,35 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent)
     return err;
 }
 
+static int byte_to_character_offset_in_line(PyObject* line, int offset) {
```
This needs to be refactored to avoid duplication, @ammaraskar

ammaraskar replied:
I did this in a hacky way, just made it an `extern` above the function. Should this be added to one of the `.h` files?

ammaraskar replied:
Oh heh this caused a build error because of `_testembed` and friends. Just moved it to `pycore_traceback.h` for now so CI passes. I'm not super familiar with how to organize Python C API functions so I'll come clean this up later once I can take a look.

pablogsal replied:
Don't worry, I will refactor it later :)

ammaraskar replied:
You should be good to go for this refactoring now.

pablogsal commented on Python/traceback.c:
```diff
@@ -501,8 +511,35 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent)
+
+#define _TRACEBACK_SOURCE_LINE_INDENT 4
+
+// TODO: Pick up filename and other stuff from the tb argument
```
We should probably fix this todo here, no?

ammaraskar replied:
@isidentical I think you added in this `TODO`, is this still something we should do?

isidentical replied:
I don't recall adding this, feel free to drop it

pablogsal commented on Python/traceback.c:
```diff
@@ -517,9 +554,52 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
+
+        start_offset = byte_to_character_offset_in_line(source_line, start_offset);
+        // Not sure why a `+ 1` is needed here for the end_offset.
+        end_offset = byte_to_character_offset_in_line(source_line, end_offset) + 1;
```
@ammaraskar I found the problem. In interval notation, the offset interval is
```
[start_offset, end_offset)
```

which is closed on the right, open on the left. As we use 1-based offsets it means that the comparison for start_offset is correct (because is inclusive) but the comparison for end_offset is not because is open.  I think that it should be:

```
        end_offset = byte_to_character_offset_in_line(source_line, end_offset - 1) + 1;
```

This is because `byte_to_character_offset_in_line` does this comparison:

```
    Py_ssize_t len = strlen(str);
    if (offset > len) {
        offset = Py_SAFE_DOWNCAST(len, Py_ssize_t, int);
    }
```

that `(offset > len)` is always going to trim valid stuff because the end_offset is always + 1 in a end line

Can you add some tests for this?

ammaraskar replied:
Yup that snippet you posted of:

```c
end_offset = byte_to_character_offset_in_line(source_line, end_offset - 1) + 1;
```

is exactly what I was thinking of. I've implemented a similar fix using `-1` to avoid the 1-indexing problem. I added some tests: https://github.com/colnotab/cpython/blob/e29545ea95c134d071a052016b010c3fd4cd83ac/Lib/test/test_traceback.py#L276-L361

But how would test this specifically/do you have suggestions for any tests for this?

pablogsal replied:
This case was failing before:

```
def foo(a: NotExistentName) -> Blech:
   ...
```

there was an off-1 error there in the highlighting. So probably we just need to add this as a test case

ammaraskar replied:
Gotcha, I did add that one so we should be good then :) https://github.com/colnotab/cpython/blob/e29545ea95c134d071a052016b010c3fd4cd83ac/Lib/test/test_traceback.py#L325-L339

isidentical committed:
bpo-43950: Print columns in tracebacks (PEP 657)

The traceback.c and traceback.py mechanisms now utilize
the newly added code.co_positions and PyCode_Addr2Location
to print carets on the specific expressions involved in a
traceback.

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>
Co-authored-by: Ammar Askar <ammar@ammaraskar.com>

isidentical commented on Python/traceback.c:
```diff
@@ -517,9 +532,54 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
+        }
+
+        Py_ssize_t start_offset = (Py_ssize_t) _PyCode_Addr2Offset(_PyFrame_GetCode(frame), code_offset);
+        Py_ssize_t end_offset = (Py_ssize_t) _PyCode_Addr2EndOffset(_PyFrame_GetCode(frame), code_offset);
```
Even though they are relatively cheap, why are we keep making `_PyFrame_GetCode` calls? (here as well as in the PyCode_Addr2Line). Also can't we just use `PyCode_Addr2Location` and get everything in one go (at worst case, it will cost 2 additional calls, but I think it worths to less complexity).

ammaraskar replied:
Changed to one `_PyFrame_GetCode` and `PyCode_Addr2Location` call.

isidentical commented on Python/traceback.c:
```diff
@@ -517,9 +532,54 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
+        }
+        if (end_offset == -1) {
+            // TODO: highlight from start_offset to the end of line
+            goto done;
```
Now that thinking, I am not sure if we should do this. Since start might be somewhere in 250, and end might be somewhere in 260 with the total line being 300 chars. So it will annotate some irrelevant code as well. @pablogsal wdyt about dropping this?

pablogsal replied:
Hummmm, not sure I follow, could you drop an example?

isidentical replied:
The `TODO: ` comment (AFAIK it was your idea) suggests that if `start_offset` is known but `end_offset` isn't (e.g start=250, end=260) we should start adding annotations from start_offset to `len(source_line)` instead of ignoring it for this case but that is wrong since if `len(source_line)` is `300` chars, then we will be annotating ~40 more chars then what we are supposed to. 


This is how it would look if we implement it:
```
long_call([245 chars], some_func_that_raises_an_error(arg), [50 chars])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

ammaraskar replied:
Yeah I think if a line is longer than 255 characters then it's probably very long to begin with and printing from somewhere in the middle to the end could be misleading. Let's get rid of this TODO and not print carets if end information isn't available.

pablogsal replied:
AH, right, I remember now. Yeah, I agree with your analysis.

isidentical commented on Python/traceback.c:
```diff
@@ -517,9 +532,54 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
+        // Convert the utf-8 byte offset to the actual character offset so we
+        // print the right number of carets. We do -1 here because the column
+        // offsets provided by _PyCode_Addr2Offset and _PyCode_Addr2EndOffset
+        // are 1-indexed, not 0-indexed.
```
Now that they are `1-indexed`, you can drop this.

isidentical commented on Python/traceback.c:
```diff
@@ -926,4 +986,3 @@ _Py_DumpTracebackThreads(int fd, PyInterpreterState *interp,
 
     return NULL;
 }
-
```
Missing newline

ammaraskar committed:
Address feedback from @isidentical

ammaraskar committed:
Remove TODO about highlighting to end

pablogsal committed:
Move byte_offset_to_character_offset to the parser API

ammaraskar committed:
Remove stale TODO

graingert commented on Lib/traceback.py:
```diff
@@ -309,7 +321,7 @@ def walk_stack(f):
     """
     if f is None:
-        f = sys._getframe().f_back.f_back
+        f = sys._getframe().f_back.f_back.f_back.f_back
```
this doesn't seem correct - 

```
Python 3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:18:27) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import traceback
>>> traceback.walk_stack(None)
<generator object walk_stack at 0x7f308f4c4700>
>>> list(traceback.walk_stack(None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/graingert/anaconda3/envs/dask-distributed-311/lib/python3.11/traceback.py", line 332, in walk_stack
    f = sys._getframe().f_back.f_back.f_back.f_back
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'f_back'
>>> 
```

pablogsal replied:
Please, don't comment on merged pull requests because they are not monitored and you risk no one looking at your comment. Always open a new issue :)
